fix: retry watch forbidden secrets

TheSpiritXIII opened the pull request:
The configuration reload fails if the caller (the Prometheus collector) does not have permission to view the secret. This change allows the configuration reload to pass, and retries the secret watch until the caller gives permission. A new unit test was added to support this change.

pintohutch reviewed (commented):
Just wanna make sure I understand the description. I think you mean:
_The configuration reload fails if the ~user~ caller (i.e. Prometheus) does not have permission to view the secret. This change allows the configuration reload to pass, and retries the secret watch until the ~user gives~ caller is given permission. A new unit test was added to support this change._

TheSpiritXIII commented on pkg/secrets/kubernetes_test.go:
```diff
@@ -617,6 +628,114 @@ func TestProvider(t *testing.T) {
+		_, err = c.CoreV1().Secrets(secret.Namespace).Update(ctx, secret, metav1.UpdateOptions{})
+		require.NoError(t, err)
+
+		requireFetchEquals(ctx, t, s, "Hello")
```
It ensures that the watch provider:
1. Does not error when calling `Add`.
2. Eventually watches the same secret from the initial `Add`.
3. Picks up that the secret changed since the initial `Add` (which comes from the Kubernetes `get` API, as that's the first call that happens after the permission is granted).
4. Picks up that the secret changed again after the last change (which comes from the Kubernetes `watch` API, as `get` is no longer called once a watch is established).

`s` is an interface that implements `Fetch` which stores data from the watch.

pintohutch replied:
Awesome - I missed that the `Secret` type of `s` is actually our internal secret, not `corev1.Secret`.

Could you add this as a comment in the test code?

TheSpiritXIII replied:
Renamed to `providerSecret`.

TheSpiritXIII commented on pkg/secrets/watch.go:
```diff
@@ -125,45 +116,74 @@ func (w *secretWatcher) secret(config *KubernetesSecretConfig) Secret {
 		defer w.mu.Unlock()
 		w.s = nil
-		return true, nil
+		return false, nil
```
`false, nil` is returned intentionally when the `context.Context` is closed. We consider this an intentional stop with no error. Unintentional stops with errors are logged. I would prefer not logging the case where we intentionally stop.

pintohutch replied:
Ok - sgtm

TheSpiritXIII commented:
@pintohutch updated the description and addressed your comments. Let me know if you disagree or want to challenge some of my comments. I'm happy to take another look!
